filesystem/azure.go

```diff
@@ -0,0 +1,196 @@
+	"io"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+)
+
+type AzureClient struct {
+	client     *storage.StorageClient
+	blobClient *storage.BlobStorageClient
+}
+
+type AzureFile struct {
+	path   string
+	logger *log.Logger
+	client *storage.BlobStorageClient
+}
+
+// convertToAzurePath function
+// like this pattern "ContainerName/BlobName"
+// Due to Azure restriction, the length of ContainerName must be 32
+func convertToAzurePath(name string) (string, string, error) {
+	afterSplit := strings.Split(name, "/")
+	if len(afterSplit) != 2 || len(afterSplit[0]) != 32 {
+		return "", "", fmt.Errorf("AzureClient : Need Correct Path Name")
+	}
+	return afterSplit[0], afterSplit[1], nil
+}
+
+// AzureClient -> Exist function
+// Only check the BlobName if exist or not
+// User should Provide corresponding ContainerName
+func (c *AzureClient) Exists(name string) (bool, error) {
+	containerName, blobName, err := convertToAzurePath(name)
+	if err != nil {
+		return false, err
+	}
+	return c.blobClient.BlobExists(containerName, blobName)
+}
+
+// AzureClient -> Rename function
+// Azure prevent user renaming their blob
+// Thus this function firstly copy the source blob,
+// when finished, delete the source blob.
+// http://stackoverflow.com/questions/3734672/azure-storage-blob-rename
+func (c *AzureClient) Rename(oldpath, newpath string) error {
+	exist, err := c.Exists(oldpath)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		return fmt.Errorf("AzureClient : oldpath doesnot exist")
+	}
+	srcContainerName, srcBlobName, err := convertToAzurePath(oldpath)
+	if err != nil {
+		return err
+	}
+	dstContainerName, dstBlobName, err := convertToAzurePath(newpath)
+	if err != nil {
+		return err
+	}
+	dstBlobUrl := c.blobClient.GetBlobUrl(dstContainerName, dstBlobName)
+	srcBlobUrl := c.blobClient.GetBlobUrl(srcContainerName, srcBlobName)
+	c.blobClient.CopyBlob(dstContainerName, dstBlobName, srcBlobUrl)
+	if dstBlobUrl != srcBlobUrl {
+		err = c.blobClient.DeleteBlob(srcContainerName, srcBlobName)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AzureClient -> OpenReadCloser function
+// implement by the providing function
+func (c *AzureClient) OpenReadCloser(name string) (io.ReadCloser, error) {
+	containerName, blobName, err := convertToAzurePath(name)
+	if err != nil {
+		return nil, err
+	}
+	return c.blobClient.GetBlob(containerName, blobName)
+}
+
+//AzureClient -> OpenWriteCloser function
+// If not exist, Create corresponding Container and blob.
+// At present, AzureFile.Write has a capacity restriction(10 * 1024 * 1024 bytes).
+// I will implent unlimited version in the future.
+func (c *AzureClient) OpenWriteCloser(name string) (io.WriteCloser, error) {
+	exist, err := c.Exists(name)
+	if err != nil {
+		return nil, err
+	}
+	containerName, blobName, err := convertToAzurePath(name)
+	if err != nil {
+		return nil, err
+	}
+	if !exist {
+		_, err := c.blobClient.CreateContainerIfNotExists(containerName, storage.ContainerAccessTypeBlob)
```

Comment:
How should I deal with the shadowing variable?
This shadowing variable represents the container if exists or not, thus I ignore it.